src/service/resources-impl.js

```diff
@@ -407,13 +410,29 @@ export class Resources {
    * @private
    */
   buildReadyResources_() {
+    // Avoid cases where elements add more elements inside of them
+    // and cause an infinite loop of building - see #3354 for details.
+    if (this.isCurrentlyBuildingPendingResources_) {
```

Comment:
Optimizers shouldn't touch it due to http://mrale.ph/blog/2014/12/24/array-length-caching.html. Confirming with closure's output:

```js
window.a=[1,2,3];window.test=function(){for(var c=window.a,b=0;b<c.length;b++)console.log(c[b])};
```

The test @mkhatib added should save us from our own changes though, like iterating in reverse.